torch/csrc/Exceptions.h

```diff
@@ -24,6 +22,45 @@ class python_error : public std::exception {};
 
 extern PyObject *THPException_FatalError;
 
+// Throwing this exception means that the python error flags have been already
+// set and control should be immediately returned to the interpreter.
+struct python_error : public std::exception {
+  python_error() : type(nullptr), value(nullptr), traceback(nullptr) {}
+
+  ~python_error() {
+    if (type || value || traceback) {
+      AutoGIL gil;
+      Py_XDECREF(type);
+      Py_XDECREF(value);
+      Py_XDECREF(traceback);
+    }
+  }
+
+  /** Saves the exception so that it can be re-thrown on a different thread */
+  inline void persist() {
+    // PyErr_Fetch overwrites the pointers
+    AutoGIL gil;
+    Py_XDECREF(type);
+    Py_XDECREF(value);
+    Py_XDECREF(traceback);
+    PyErr_Fetch(&type, &value, &traceback);
+  }
+
+  /** Sets the current Python error from this exception */
+  inline void restore() {
+    // PyErr_Restore steals references
+    AutoGIL gil;
+    Py_XINCREF(type);
+    Py_XINCREF(value);
+    Py_XINCREF(traceback);
+    PyErr_Restore(type, value, traceback);
```

Comment:
Then the python_error can only be restored once. That makes the API more complicated.